File: go/2487.go
package main

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func removeNodes(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}
	ms := make([]*ListNode, 0)
	cur := head
	for cur != nil {
		for len(ms) > 0 && cur.Val > ms[len(ms)-1].Val {
			ms = ms[:len(ms)-1]
		}
		ms = append(ms, cur)
		cur = cur.Next
	}
	for i := 0; i < len(ms)-1; i++ {
		ms[i].Next = ms[i+1]
	}
	ms[len(ms)-1].Next = nil
	return ms[0]
}

File: cpp/2487.cpp
#include <vector>

// Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  ListNode *removeNodes(ListNode *head) {
    if (!head)
      return nullptr;
    std::vector<ListNode *> ms; // monotonic stack
    ListNode *cur = head;

    while (cur) {

      // if the property will be break if cur node push into
      while (ms.size() && cur->val > ms.back()->val)
        ms.pop_back();

      ms.push_back(cur);
      cur = cur->next;
    }
    // link them
    for (size_t i = 0; i < ms.size() - 1; i++)
      ms[i]->next = ms[i + 1];
    ms.back()->next = nullptr;
    return ms[0];
  }
};
